source/common/buffer/buffer_impl.h

```diff
@@ -297,13 +303,44 @@ class Slice {
     return num_pages * PageSize;
   }
 
+  static StoragePtr newStorage(uint64_t capacity) {
+    ASSERT(sliceSize(default_slice_size_) == default_slice_size_,
+           "default_slice_size_ incompatible with sliceSize()");
+
+    StoragePtr storage;
+    if (capacity == default_slice_size_ && !free_list_.empty()) {
+      storage = std::move(free_list_.back());
+      free_list_.pop_back();
+    } else {
+      storage.reset(new uint8_t[capacity]);
+    }
+
+    return storage;
+  }
+
+  static void freeStorage(StoragePtr storage, uint64_t capacity) {
+    if (storage == nullptr) {
+      return;
+    }
+
+    if (capacity == default_slice_size_ && free_list_.size() < free_list_max_) {
```

Comment:
Potentially, but then we'd lose ASAN coverage of whether the freelist is working properly, eg some case where the slice is both free'd and put in the freelist, etc.

Is there any way to tell ASAN to consider use of the freelist like a malloc and free?

Reply:
ping.  Thoughts on removing this freelist.

Reply:
I ran the buffer speedtest both ways, and it was about 10% slower without the freelist.

Reply:
I just ran buffer_speed_test on my linux workstation and I think I see the benchmark being faster when the freelists are disabled.  Not sure what's different between our runs.

Reply:
Here's what I got. Note that this is with the change I'm about to push, which reduces the number of allocations in a Reservation from 8 to 1 (and I think makes the interface slightly cleaner too). 

with freelist:
```
----------------------------------------------------------------------------
Benchmark                                  Time             CPU   Iterations
----------------------------------------------------------------------------
bufferReserveCommit/131072               410 ns          410 ns    102819143
bufferReserveCommitPartial/131072        369 ns          369 ns    114201963
```

without freelist:
```
----------------------------------------------------------------------------
Benchmark                                  Time             CPU   Iterations
----------------------------------------------------------------------------
bufferReserveCommit/131072               500 ns          500 ns     84142824
bufferReserveCommitPartial/131072        454 ns          454 ns     92758817
```

Reply:
I don't really understand why `bufferReserveCommit` is so much slower with the freelist. It should be mostly unused in that case (all slices go from reservation into the buffer), except the periodic drain when > MaxBufferLength.

Reply:
The extra cost seems associated with use of the thread_local qualifier.  Removing thread_local brings runtime of the freelist version in line with the no-freelist version; the not-threadlocal static freelist version seems as fast as the no freelist version.

One potential improvement would be to make some adjustments so that a reference to the thread local is accessed once outside the loop that allocates the 8 buffer slices from the freelist.

Reply:
I tried pulling the thread_local lookup out of the loop, but perf looks nearly identical. Doesn't tcmalloc have to do the same thing (access thread_local data) internally for the non-freelist version (using tcmalloc for the freelist)?

Reply:
by default tcmalloc uses per-cpu caches, not per thread.  It doesn't end up using the thread_local c++ keyword.